src/core/qgscoordinatereferencesystem_p.h

```diff
@@ -124,6 +136,9 @@ class QgsCoordinateReferenceSystemPrivate : public QSharedData
     bool mIsValid = false;
 
 #if PROJ_VERSION_MAJOR>=6
+
+    // this is the "master" proj object, to be used as a template for new proj objects created on different threads ONLY.
+    // Always use threadLocalProjObject() instead of this.
```

Comment:
are we sure that mPj will be destroyed before the PJ_CONTEXT* that owns it ?

Reply:
This won't have changed in this PR, but I could add mPj to removeObjectsBelongingToCurrentThread for safety?

Reply:
@rouault can you get the context which a PJ* belongs to? Does api exist for that?

Reply:
> This won't have changed in this PR

agreed

> but I could add mPj to removeObjectsBelongingToCurrentThread for safety?

then QgsCoordinateReferenceSystemPrivate will have a nullptr mPj and be inconsistent, no ? (in the unlikely scenario that a temporary thread would create a CRS and pass it to the main thread)
But, assuming that a QgsCoordinateReferenceSystem object cannot be simultaneously used by several threads at the same time, I'm wondering if instead of that logic of having a per-thread cache of PJ* objects , you couldn't just use proj_assign_context() with the current TLS context ? And at destruction time of mPj, you create a temporary PJ_CONTEXT like done in QgsCoordinateTransformPrivate::freeProj()

Reply:
> can you get the context which a PJ* belongs to?

I don't think so. But proj_assign_context() should be cheap enough to do it systematically

Reply:
> then QgsCoordinateReferenceSystemPrivate will have a nullptr mPj and be inconsistent, no 

That's fine -- there's already circumstances when this can be nullptr, so it's already guarded against this.

> But, assuming that a QgsCoordinateReferenceSystem object cannot be simultaneously used by several threads at the same time

it can be -- that's the root cause of the issue here.

> And at destruction time of mPj, you create a temporary PJ_CONTEXT like done in QgsCoordinateTransformPrivate::freeProj()

Will do - incoming shortly.

Reply:
> I don't think so. 

No worries - I can easily workaround this by storing the owner PJ_CONTEXT in QgsCoordinateReferenceSystemPrivate